### PR TITLE
add messages regarding random effects grouping factors

### PR DESCRIPTION
**Pull request by FBartos**
fixes 2/3 of: https://github.com/jasp-stats/jasp-test-release/issues/806

1) sets an error on the empty default table if no random effects grouping factor is selected to notify the user about it.
image

2) adds a note with the random effects grouping factors specification
image

### CODE REVIEW

**Comment by AlexanderLyNL** [comment]
Sorry, I should have made this clear before, but please don't put a big red box there, because that feels like a punishment. Instead, show an empty table with that message in the footnote, and the fixed effects labels, but with ... in the output, so it's clear to the user that the analysis is waiting for something.

**Comment by FBartos** [comment]
changing 1) to a note instead
image

**Comment by AlexanderLyNL** [comment]
- Yes that looks great. To make it even more obvious, I'd have contGamma on the left, or what ever variables entered in "contGamma" and then dots in the output. 
- The solution with the note is great, the only problem I now have is that it's not that obvious, whereas the previous red box was too obvious
- @TimKDJ what would be good user feedback here? A highlighted "Random effects grouping factors box" <- I noticed something happening with this type of user feedback via the upgrader? Or the note text in green? Or instead of dots in the table the message "Please also specify 'random effects' " but not as a big red box?

**Comment by FBartos** [comment]
So, this would be the output with the dots
image

And I also fixed when "Number of trials" was not correctly enabled when "Binomial (aggregated)" was selected.

**Comment by AlexanderLyNL** [comment]
- I think that a plug-in of "contGamma" under effects (and all other variables names specified under fixed effects variables) could be helpful. 
-  The reason is that it provides the user with feedback acknowledging the user "I understand that you added contGamma" and the dot output conveys that it's waiting for something. @TimKDJ is that still the rule? Or are we going for empty rows now?
- In principle the note already informs the user the reason why it's not producing any results yet
- It would have been easier if random effects were reported, then you can have a random effects table with a column "Random effects" and under it ... stating that the analysis needs a random effect. 
- The note takes over the role of the previous point.
- Again, we need something more prominent than the note, especially for those with old eyes. But that's a different issue. 

An alternative approach would be to change the UI:
- enter all the variables into "Covariates" (instead of fixed effects): Produce standard GLM result
- Section called "Random effects" takes in subset of variables which can be selected and specified as random effects
- The idea is loosely based on the UI of linear regression, where we specify nuisance parameters under the "Model" section

**Comment by EJWagenmakers** [comment]
Without having given this specific suggestion much thought (Frantisek, Lexi, let me know if you want to discuss this in a zoom meeting or something!) I think we should at least give Frantisek/Henrik's GUI a go for 0.13 -- we are in the final, final stages of getting out a new version and unless there is a blatant error I think we should avoid making structural changes. We can always change the GUI for the next version.

**Comment by TimKDJ** [comment]
I think for now a footnote should suffice. But maybe it should read `This analysis requires at least one random effects grouping factor to run`. At that point everyone should be able to work out why the analysis is in a waiting state.

The long term solution is to make the analysis work without the grouping factor as described in [this issue](https://github.com/jasp-stats/INTERNAL-jasp/issues/932); it's obviously better to make the analysis do what a user would reasonably expect it to :)

**Comment by FBartos** [comment]
So, I changed the message as suggested (it's shown only if the Random-effects grouping factor is missing):
image

and added a message to address this issue (https://github.com/jasp-stats/jasp-test-release/issues/836)- it's not clear that one needs to specify the "Number of trials" variable in order to run the analysis with "Binomial (aggregated)" family 
image

I also fixed one issue that would prevent the BGLMM with "Binomial (aggregated)" family to run

File: JASP-Engine/JASP/R/MixedModelsCommon.R [hunk 1]
```diff
@@ -552,7 +551,17 @@
...
     
     if (is.null(model)) {
-      ANOVAsummary$setExpectedSize(1)
+      if (options$dependentVariable != "" &&
+          length(options$fixedVariables) > 0 &&
+          length(options$randomVariables) == 0) {
+        ANOVAsummary$addFootnote(.mmMessageMissingRE)
+      }
+      if (type == "GLMM") {
+        if (options$family == "binomial_agg" &
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/MixedModelsCommon.R]: `&&`

File: JASP-Engine/JASP/R/MixedModelsCommon.R [hunk 2]
```diff
@@ -2904,6 +2927,18 @@
...
 }
+.mmMessageREgrouping    <- function(RE_grouping_factors) {
+  gettextf(
+    "The following %s used %s random effects grouping %s: %s.",
+    ifelse(length(RE_grouping_factors) < 2, "variable is", "variables are"),
+    ifelse(length(RE_grouping_factors) < 2, "as a", "as"),
+    ifelse(length(RE_grouping_factors) < 2, "factor", "factors"),
+    ifelse(length(RE_grouping_factors) == 0, "...", paste0("'", RE_grouping_factors, "'", collapse = ", "))
+  )
+}
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/MixedModelsCommon.R]: Use `sprintf(ngettext())` for this. Aside from the strings not being in `gettext()` the structure of this sentence might need to be entirely different in another language.

**Review by TimKDJ — CHANGES_REQUESTED** [review]

File: JASP-Engine/JASP/R/MixedModelsCommon.R (hunk 2 above)
Comment (FBartos) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: Yeah, thanks. I keep forgetting those.

What do you advise to do about it then?

Comment (FBartos) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: I could maybe create two complete sentences for each case

Comment (TimKDJ) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: Something like:
```
sprintf(ngettext(length(RE_grouping_factors) < 2,
                 "The following variable is used as random effects grouping factor: %s.",
                 "The following variables are used as random effects grouping factors: %s."),
        ifelse(length(RE_grouping_factors) == 0, "...", paste0("'", RE_grouping_factors, "'", collapse = ", ")))
```

untested though

File: JASP-Engine/JASP/R/MixedModelsCommon.R [hunk 3]
```diff
@@ -2986,32 +2968,25 @@
...
-  if (value < 2) {
-    return(gettextf("%i observation was removed due to missing values.",
-                    value))
-  } else{
-    return(gettextf("%i observations were removed due to missing values.",
-                    value))
-  }
+  sprintf(
+    ngettext(
+      value < 2,
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/MixedModelsCommon.R]: Unless you specifically want to group `0` and `1` together you should use `ngettext(length(value), "..", "..")` (so no boolean, but an integer as argument to `ngettext()`)

`0 observation was removed due to missing values.` is not correct.

This goes for quite a few of these changed lines

**Review by TimKDJ — CHANGES_REQUESTED** [review]

File: JASP-Engine/JASP/R/MixedModelsCommon.R (hunk 3 above)
Comment (FBartos) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: the 's' is actually correct at the end - there might be more missing values within one observation. We do not check those.

These messages aren't called when the value == 0

Comment (TimKDJ) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: We've solved this over Slack, but:

`0 observation was removed` is correct not though :)
These are your sentences:
```
 "%i observation was removed due to missing values.",
 "%i observations were removed due to missing values."
```

Comment (FBartos) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: 0 observation was removed cannot happen. The function is called only if `value > 0`

Comment (TimKDJ) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: I know, but the English is not correct. It was in response to this:

> the 's' is actually correct at the end - there might be more missing values within one observation. We do not check those.

Comment (FBartos) [reply JASP-Engine/JASP/R/MixedModelsCommon.R]: oh, my English is probably not good enough to understand the problem with my English :D

**Review by TimKDJ — APPROVED** [review]

**Pull request merged**